src/shrinkage.jl

```diff
@@ -1,91 +1,46 @@
-    for (j, re) in enumerate(m.reterms)
-        cnms = re.cnames
-        cnms ⊆ fenms || throw(ArgumentError("No corresponding fixed effect for random effects $(setdiff(cnms, fenms)): there is no estimated grand mean to measure shrinkage towards"))
-        # XXX Should m.X return a view into m.Xymat?
-        Xmat = view(m.X, :, [findfirst(==(nm), fenms) for nm in cnms])
-        levs = re.levels
-        refs = re.refs
-        grpest = Matrix{Union{Missing,T}}(missing, (length(levs), length(cnms)))
-        for i in eachindex(levs)
-            rows = findall(==(i), refs)
-            try
-                grpest[i, :] = view(Xmat, rows, :) \ view(yvec, rows)
-            catch
-            end
-        end
-        push!(cvec, CoefByGroup(cnms, levs, Xmat \ yvec, ranefs[j]', grpest))
+function shrinkageplot(
+    m::LinearMixedModel{T},
+    gf::Symbol=first(fnames(m)),
+    θref::AbstractVector{T}=10000 .* m.optsum.initial,
+) where {T}
+    reind = findfirst(==(gf), fnames(m))  # convert the symbol gf to an index
+    if isnothing(reind)
+        throw(ArgumentError("gf=$gf is not one of the grouping factor names, $(fnames(m))"))
     end
-    NamedTuple{fnames(m)}((cvec...,))
-end
-
-shrinkageplot(m::LinearMixedModel, gf::Symbol=first(fnames(m))) = shrinkageplot(shrinkage(m)[gf])
-
-shrinkageplot(cg::CoefByGroup) = shrinkageplot!(Figure(resolution=(1000,1000)), cg)
-
-function shrinkageplot!(f::Figure, cg::CoefByGroup)
-    shrinkage2d!(Axis(f[1,1]), cg)
-    f
-end
-
-function shrinkage2d!(a::Axis, cg::CoefByGroup{T}, inds=(1, 2)) where T
-    i, j = inds
-    fxd = cg.fixed
-    scatter!(a, view(fxd, i:i), view(fxd, j:j), color=:green, label="Pop")
-    u = view(cg.condmodes, :, i) .+ fxd[i]
-    v = view(cg.condmodes, :, j) .+ fxd[j]
-    scatter!(a, u, v, color=:blue, label="Mixed")
-    x = view(cg.grpest, :, i)
-    y = view(cg.grpest, :, j)
-    missgrp = ismissing.(x) .| ismissing.(y)
-    if any(missgrp)
-        nonmiss = .!(missgrp)
-        x = convert(Vector{T}, view(x, nonmiss))
-        y = convert(Vector{T}, view(y, nonmiss))
-        u = view(u, nonmiss)
-        v = view(v, nonmiss)
-    else
-        x = convert(Vector{T}, x)
-        y = convert(Vector{T}, y)
+    reest = ranef(m)[reind]               # random effects conditional means at estimated θ
+    reref = ranef(updateL!(setθ!(m, θref)))[reind]  # same at θref
+    updateL!(setθ!(m, m.optsum.final))    # restore parameter estimates and update m
+    cnms = m.reterms[reind].cnames
+    f = Figure(; resolution=(1000, 1000)) # use an aspect ratio of 1 for the whole figure
+    k = size(reest, 1)  # dimension of the random-effects vector per level of gf
+    for i in 2:k                          # strict lower triangle of panels
+        for j in 1:(i - 1)
+            ax = Axis(f[i - 1, j])
+            x, y = view(reref, j, :), view(reref, i, :)
+            scatter!(ax, x, y; color=(:red, 0.25))   # reference points
```

Comment:
I haven't gotten a sense of a canonical Makie interface approach yet.  In R you had to do all of the drawing in one pass because the R graphics model is based on that from S which assumed a pen plotter.  This led to functions like lattice::xyplot that have an absurd number of optional arguments.

Makie feels a bit different in that you can create the layout and later make modifications to it and everything will get sorted out when the figure is rendered.  So I am not sure if the customizing should be done in this function via optional arguments or later in a modifier function.

I agree that it should be addressed but I think we will be okay to let it marinate until we know the system better.

Reply:
@deepayan pointed out to me that an important difference between lattice and ggplot2 is the use of a panel function in lattice.  In lattice you can separate to some extent the layout of the panels separately from the contents that will be displayed in them.   You have the advantage in R of being able to examine the state inside the calling function from within the panel function.  Here it may be possible to do something similar by passing `ax`, `i` and `j` to a panel function that is a closure.

Reply:
OOOh, I like that idea.